source/extensions/filters/http/rate_limit_quota/filter.h

```diff
@@ -0,0 +1,145 @@
+      : setting_(std::move(settings)) {}
+
+  absl::StatusOr<BucketId> generateBucketId(const Http::Matching::HttpMatchingDataImpl& data,
+                                            Server::Configuration::FactoryContext& factory_context,
+                                            RateLimitQuotaValidationVisitor& visitor) const;
+
+private:
+  envoy::extensions::filters::http::rate_limit_quota::v3::RateLimitQuotaBucketSettings setting_;
+};
+
+class RateLimitOnMactchActionFactory
+    : public Matcher::ActionFactory<RateLimitOnMactchActionContext> {
+public:
+  std::string name() const override { return "rate_limit_quota"; }
+
+  Matcher::ActionFactoryCb
+  createActionFactoryCb(const Protobuf::Message& config, RateLimitOnMactchActionContext&,
+                        ProtobufMessage::ValidationVisitor& validation_visitor) override {
+    // Validate and then retrieve the bucket settings from config.
+    const auto bucket_settings =
+        MessageUtil::downcastAndValidate<const envoy::extensions::filters::http::rate_limit_quota::
+                                             v3::RateLimitQuotaBucketSettings&>(config,
+                                                                                validation_visitor);
+    return [bucket_settings = std::move(bucket_settings)]() {
+      return std::make_unique<RateLimitOnMactchAction>(std::move(bucket_settings));
+    };
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<
+        envoy::extensions::filters::http::rate_limit_quota::v3::RateLimitQuotaBucketSettings>();
+  }
+};
+
+class RateLimitQuotaFilter : public Http::PassThroughFilter,
+                             public RateLimitQuotaCallbacks,
+                             public Logger::Loggable<Logger::Id::filter> {
+public:
+  RateLimitQuotaFilter(FilterConfigConstSharedPtr config,
+                       Server::Configuration::FactoryContext& factory_context,
+                       RateLimitClientPtr client)
+      : config_(std::move(config)), rate_limit_client_(std::move(client)),
+        factory_context_(factory_context) {
+    createMatcher();
+  }
+
+  // Http::PassThroughDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override;
+  void onDestroy() override;
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override;
+
+  // RateLimitQuota::RateLimitQuotaCallbacks
+  void onReceive(envoy::service::rate_limit_quota::v3::RateLimitQuotaResponse*) override {}
+
+  // Perform request matching. It returns the generated bucket ids if the matching succeeded and
+  // returns the error status otherwise.
+  absl::StatusOr<BucketId> requestMatching(const Http::RequestHeaderMap& headers);
+
+  ~RateLimitQuotaFilter() override = default;
+
+private:
+  // Create the matcher factory and matcher.
+  void createMatcher();
+
+  FilterConfigConstSharedPtr config_;
+  RateLimitClientPtr rate_limit_client_;
```

Comment:
It does not seem to be conceptually correct. Filter should share the thread-local state (sort of cache) of quotas and this state includes the client that gets the bucket assignments.